### PR TITLE
perf(sql): use radix sort to speed up single integer column order by

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Introduces a quick or radix sort-based factory used for unlimited ORDER BY over a single INT / IPv4 / LONG / TIMESTAMP / DATE column or expression. The memory overhead is similar to the existing `SortedRecordCursorFactory`: each row takes 32 bytes vs 36 bytes in the red-black tree-based factory (the latter assumes #4888).

### Benchmarks

The following numbers were obtained on i7-1185G7 CPU with 32GB RAM, running Ubuntu 22.04 64-bit and OpenJDK 17.0.12. The dataset is the one from ClickBench.

```sql
-- 99M rows
-- master: 320s
-- radix sort: 4s
SELECT * FROM hits ORDER BY UserID;

-- 10M rows
-- master: 14s
-- radix sort: 400ms
SELECT * FROM (hits LIMIT 10000000) ORDER BY UserID;

-- 1M rows
-- master: 456ms
-- radix sort: 50ms
SELECT * FROM (hits LIMIT 1000000) ORDER BY UserID;

-- 10K rows
-- master: 12ms
-- radix sort: 9ms
SELECT * FROM (hits LIMIT 10000) ORDER BY UserID;

-- 100 rows
-- master: 5ms
-- radix sort: 3ms
SELECT * FROM (hits LIMIT 100) ORDER BY UserID;
```

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 125 / 134 (93.28%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java|26|31|83.87%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursor.java|82|86|95.35%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 131 / 140 (93.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java|27|32|84.38%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursor.java|86|90|95.56%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|2|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor baseCursor = base.getCursor(executionContext);
+        try {
+            cursor.of(baseCursor, executionContext);
+            return cursor;
+        } catch (Throwable ex) {
+            baseCursor.close();
+            cursor.close();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java:93]: this is ripe for duble-free. We must enforce predictable behavior out of "of" method. As in, it must be responsible for closing the "baseCursor". The problem exists in other classes that call "of"

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java:92]: Good spot. Fixed it in this and other places in 383831d31507a93c7f83849e85f0e82dcd46a646

File: core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                configuration,
+                columnIndex,
+                metadata.getColumnType(columnIndex),
+                sortColumnFilter.getColumnIndex(0) > 0
+        );
+        this.sortColumnFilter = sortColumnFilter;
+    }
+
+    public static void addSortKeys(PlanSink sink, ListColumnFilter filter) {
+        sink.attr("keys").val('[');
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java:67]: there seem to be code duplication here

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java:67]: Not sure I follow. What kind of code duplication are you referring to?

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java]: I've noticed the duplication now. Removed it in 6f911157d7d06a5dd30bf38f543ba86b1f5c15ba

Interestingly, my IDEA CE didn't highlight this method.

File: core/src/main/java/io/questdb/griffin/engine/orderby/SortedRecordCursor.java [hunk 3]
```diff
@@ -82,14 +82,13 @@ public SymbolTable newSymbolTable(int columnIndex) {
     }
 
     @Override
-    public void of(RecordCursor base, SqlExecutionContext executionContext) {
+    public void of(RecordCursor baseCursor, SqlExecutionContext executionContext) {
+        this.baseCursor = baseCursor;
+        chainCursor = chain.getCursor(baseCursor);
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/orderby/SortedRecordCursor.java]: Yes, it's safe to do - we only access `root` and the code checks for `-1` value. However, this doesn't look nice, so I changed the base cursor to be managed by `SortedRecordCursor`, not `RecordTreeChain.TreeCursor` here: 749e46693b91e71770f79fd608048552f0494a86

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 222 / 266 (83.46%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SelectedRecordCursorFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/CountRecordCursorFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|7|28.57%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|11|45.45%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/LimitRecordCursorFactory.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByRecordCursorFactory.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/join/RecordAsAFieldRecordCursorFactory.java|8|11|72.73%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizePartiallySortedLightRecordCursor.java|9|11|81.82%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByRecordCursorFactory.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java|23|27|85.19%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongSortedLightRecordCursor.java|86|90|95.56%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByNotKeyedRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/SelectedRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|2|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**